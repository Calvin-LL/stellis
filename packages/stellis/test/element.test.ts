import { describe, it, expect } from 'vitest';
import transformStellis from './utils';

describe('elements', () => {
  it('should transform basic', async () => {
    const code = `
      <h1>Hello World</h1>
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform with static attribute', async () => {
    const code = `
      <h1 title="hello world">Hello World</h1>
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform with dynamic attribute', async () => {
    const code = `
      <h1 title={getTitle()}>Hello World</h1>
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform with async dynamic attribute', async () => {
    const code = `
      <h1 title={await getTitle()}>Hello World</h1>
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform with static child element', async () => {
    const code = `
      <div>
        <h1>Hello World</h1>
        <h1>Hello World</h1>
      </div>
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform with dynamic child element', async () => {
    const code = `
      <div>
        {dynamicChild()}
        {dynamicChild()}
      </div>
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform with dynamic child element', async () => {
    const code = `
      <div>
        {await dynamicChild()}
        {await dynamicChild()}
      </div>
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform with nested static child element', async () => {
    const code = `
      <div>
        <div>
          <h1>Hello World</h1>
          <h1>Hello World</h1>
        </div>
        <div>
          <h1>Hello World</h1>
          <h1>Hello World</h1>
        </div>
      </div>
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform class attribute', async () => {
    const code = `
      <h1 class="example">Hello</h1>;
      <h1 class={["a", condB && b]}>Array</h1>;
      <h1 class={{ a: true, b: condB, c: condC }}>Object</h1>;
      <h1 class={["a", { b: condB }, [condC && "c"]]}>Nested</h1>;
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform class:* attribute', async () => {
    const code = `
      <h1 class:example>Hello</h1>;
      <h1 class:a class:b={cond}>Another Example</h1>;
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform style attribute', async () => {
    const code = `
      <div style={{ color: 'red' }}>Hello World</div>;
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform style:* attribute', async () => {
    const code = `
      <h1 style:color="red">Red Heading 2</h1>;
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
  it('should transform set:html attribute', async () => {
    const code = `
      <div set:html="<script>Hello World</script>" />
    `;

    expect(await transformStellis(code)).toMatchSnapshot();
  });
});
