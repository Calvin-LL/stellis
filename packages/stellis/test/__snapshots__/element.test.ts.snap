// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`elements > should transform basic 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
const _template = \\"<h1>Hello World</h1>\\";
_$$html(_template, true);"
`;

exports[`elements > should transform class attribute 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
import { $$class as _$$class } from \\"stellis\\";
const _template = [\\"<h1 \\", \\">Hello</h1>\\"],
  _template2 = [\\"<h1 \\", \\">Array</h1>\\"],
  _template3 = [\\"<h1 \\", \\">Object</h1>\\"],
  _template4 = [\\"<h1 \\", \\">Nested</h1>\\"];
_$$html(_template, true, _$$class(\\"example\\"));
_$$html(_template2, true, _$$class([\\"a\\", condB && b]));
_$$html(_template3, true, _$$class({
  a: true,
  b: condB,
  c: condC
}));
_$$html(_template4, true, _$$class([\\"a\\", {
  b: condB
}, [condC && \\"c\\"]]));"
`;

exports[`elements > should transform class:* attribute 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
import { $$class as _$$class } from \\"stellis\\";
const _template = [\\"<h1 \\", \\">Hello</h1>\\"],
  _template2 = [\\"<h1 \\", \\">Another Example</h1>\\"];
_$$html(_template, true, _$$class(\\"example\\"));
_$$html(_template2, true, _$$class(\\"a\\", cond && \\"b\\"));"
`;

exports[`elements > should transform set:html attribute 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
const _template = \\"<div><script>Hello World</script></div>\\";
_$$html(_template, false);"
`;

exports[`elements > should transform style attribute 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
import { $$style as _$$style } from \\"stellis\\";
const _template = [\\"<div \\", \\">Hello World</div>\\"];
_$$html(_template, true, _$$style({
  color: 'red'
}));"
`;

exports[`elements > should transform style:* attribute 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
import { $$style as _$$style } from \\"stellis\\";
const _template = [\\"<h1 \\", \\">Red Heading 2</h1>\\"];
_$$html(_template, true, _$$style({
  color: \\"red\\"
}));"
`;

exports[`elements > should transform with async dynamic attribute 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
import { $$attr as _$$attr } from \\"stellis\\";
const _template = [\\"<h1 \\", \\">Hello World</h1>\\"];
async () => {
  const _v = await getTitle();
  return _$$html(_template, true, _$$attr(\\"title\\", _v));
};"
`;

exports[`elements > should transform with dynamic attribute 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
import { $$attr as _$$attr } from \\"stellis\\";
const _template = [\\"<h1 \\", \\">Hello World</h1>\\"];
_$$html(_template, true, _$$attr(\\"title\\", getTitle()));"
`;

exports[`elements > should transform with dynamic child element 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
const _template = [\\"<div>\\", \\"</div>\\"];
_$$html(_template, true, () => [dynamicChild(), dynamicChild()]);"
`;

exports[`elements > should transform with dynamic child element 2`] = `
"import { $$html as _$$html } from \\"stellis\\";
const _template = [\\"<div>\\", \\"</div>\\"];
_$$html(_template, true, [async () => await dynamicChild(), async () => await dynamicChild()]);"
`;

exports[`elements > should transform with nested static child element 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
const _template = [\\"<div>\\", \\"</div>\\"],
  _template2 = [\\"<div>\\", \\"</div>\\"],
  _template3 = \\"<h1>Hello World</h1>\\",
  _template4 = \\"<h1>Hello World</h1>\\",
  _template5 = [\\"<div>\\", \\"</div>\\"],
  _template6 = \\"<h1>Hello World</h1>\\",
  _template7 = \\"<h1>Hello World</h1>\\";
_$$html(_template, true, [_$$html(_template2, true, [_$$html(_template3, true), _$$html(_template4, true)]), _$$html(_template5, true, [_$$html(_template6, true), _$$html(_template7, true)])]);"
`;

exports[`elements > should transform with static attribute 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
const _template = \\"<h1 title=\\\\\\"hello world\\\\\\">Hello World</h1>\\";
_$$html(_template, true);"
`;

exports[`elements > should transform with static child element 1`] = `
"import { $$html as _$$html } from \\"stellis\\";
const _template = [\\"<div>\\", \\"</div>\\"],
  _template2 = \\"<h1>Hello World</h1>\\",
  _template3 = \\"<h1>Hello World</h1>\\";
_$$html(_template, true, [_$$html(_template2, true), _$$html(_template3, true)]);"
`;
